### PR TITLE
[WIP] virtctl image-upload: get upload proxy URL from route/ingress

### PR DESCRIPTION
**Pull request by igoihman**
**What this PR does / why we need it**:
Add auto discovery of cdi-uploadproxy host, based on the platform we are running (k8s or openshift).
cdi-uploadproxy URL is now optional, if it's not present we'll try to detect it automatically and fail if there  is no Route/Ingress for uploadproxy service.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
virtctl upload auto-detect upload proxy URL
```

### CODE REVIEW

**Comment by gbenhaim** [comment]
ci test please

File: pkg/virtctl/imageupload/imageupload.go [hunk 1]
```diff
@@ -373,3 +382,92 @@ func getAndValidateUploadPVC(client kubernetes.Interface, namespace, name string
...
 	return pvc, nil
 }
+
+func findUploadHost(virtClient kubecli.KubevirtClient, namespace string) (string, error) {
+	isOpenshift, err := isOpenshift(virtClient)
+	if err != nil {
+		return "", err
+	}
+	if isOpenshift {
+		return getHostFromOpenshiftRoute(virtClient, namespace)
```
[the remark this answers is not in the document]
Comment (igoihman) [reply pkg/virtctl/imageupload/imageupload.go]: done

File: pkg/virtctl/imageupload/imageupload.go [hunk 2]
```diff
@@ -172,6 +174,13 @@ func (c *command) run(cmd *cobra.Command, args []string) error {
 	if err != nil {
 		return err
 	}
+
+	if uploadProxyURL == "" {
+		uploadProxyURL, err = findUploadHost(virtClient, "kube-system")
```
[the remark this answers is not in the document]
Comment (igoihman) [reply pkg/virtctl/imageupload/imageupload.go]: done, I'm searching for ingress/route in all namespaces now, taking the host from the first one I get

File: pkg/virtctl/imageupload/imageupload.go [hunk 3]
```diff
@@ -373,3 +382,92 @@ func getAndValidateUploadPVC(client kubernetes.Interface, namespace, name string
...
+	if err != nil {
+		return "", err
+	}
+	if isOpenshift {
+		return getHostFromOpenshiftRoute(virtClient, namespace)
+	}
+	return getHostFromIngress(virtClient, namespace)
+}
+
+func getHostFromIngress(client kubernetes.Interface, namespace string) (string, error) {
```
[the remark this answers is not in the document]
Comment (igoihman) [reply pkg/virtctl/imageupload/imageupload.go]: done

File: pkg/virtctl/imageupload/imageupload.go [hunk 4]
```diff
@@ -373,3 +382,92 @@ func getAndValidateUploadPVC(client kubernetes.Interface, namespace, name string
...
+				if path.Backend.ServiceName == uploadProxyServiceName {
+					return rule.Host, nil
+				}
+			}
+		}
+	}
+	return "", fmt.Errorf("No Ingress for service %s was found", uploadProxyServiceName)
+}
+
+func extractHostFromRoute(data []byte) (string, error) {
```
[the remark this answers is not in the document]
Comment (igoihman) [reply pkg/virtctl/imageupload/imageupload.go]: done

**Comment by igoihman** [comment]
@rmohr The intention is to support easy upload,  Ingress object can be created in any namespace. This way the user won't have to specify upload proxy URL every time he wants to upload an image using virtctl and will have an FQDN that exposes cdi-uploadproxy service.

**Comment by rmohr** [comment]
@igoihman so you expect that every user creates its own upload ingress route?

**Comment by igoihman** [comment]
@rmohr ingress/route should be created once per setup in order to expose cdi-uploadproxy service.
Until now, the user had to do port forwarding or other manipulations in order to access the proxy and give a valid URL. This way, he won't have to do it and only create an ingress, the discovery will happen during the run.
Same user that had permissions for port forwarding can now create an ingress instead.

**Comment by rmohr** [comment]
> @rmohr ingress/route should be created once per setup in order to expose cdi-uploadproxy service.

So different users only see different namespaces. The admin creates the ingress/service once somewhere. He would  now need to make sure that all users can see that route, right?

> Until now, the user had to do port forwarding or other manipulations in order to access the proxy and give a valid URL. This way, he won't have to do it and only create an ingress, the discovery will happen during the run.

That seems to be more a "developer" workaround than a realistic use. I would have expected that an admin creates the "ingress" route once and then somehow informs users where they can upload. Only this last step needs to be discoverable somehow.

> Same user that had permissions for port forwarding can now create an ingress instead.

Normally all users can do port-forward in namespaces they own (do you mean host-port?).


This solution seems to make things worse than before. I would rather consider doing something like this (very rough):

 * Create in the CDI namespace an ingress route and add a special label/annotation on it
 * I don't think that ingress objects contain too sensitive data, but I would not make it visible to arbitrary users, so let a controller back-propagate the FQDN to a config-map in the cdi namespace
 * As part of the CDI setup make that config-map readable for every user in the cluster
 * let kubectl lookup this config map

The only thing which a user needs to know now is the CDI install namespace. I think it is save to assume that it it is "cdi" by default.

**Comment by rmohr** [comment]
Also something like a `non-namespaced` `CDIUserConfig` Custom Resource could be used. Then the namespace variable is removed too.

**Comment by mhenriks** [comment]
Really great ideas @rmohr!  The upload controller can watch Ingress/Routes in the cdi namespace and create the CR.

**Comment by aglitke** [comment]
I also think this sounds like a great plan.  The CR approach (as opposed to the ConfigMap) seems to be the best approach for ease of use.

**Comment by rmohr** [comment]
One question: Don't you alrady have a custom  resource for fetching access tokens? Could you return there the proxy too?

**Comment by rmohr** [comment]
/close

As agreed above this will be superseded with another approach.

**Pull request closed without merging**